### PR TITLE
Sea Turtules: Danielle Whyte - Task List API

### PR DESCRIPTION
**Pull request by daniellewhyte**

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 
 db = SQLAlchemy()
-migrate = Migrate()
+migrate = Migrate(compare_type=True)
```
Comment (kelsey-steven-ada) [inline app/__init__.py:9]: Interesting, what did you run into that caused a need to pass `compare_type=True` to the Migrate constructor?

File: app/__init__.py [hunk 2]
```diff
@@ -30,5 +30,10 @@ def create_app(test_config=None):
     migrate.init_app(app, db)
 
     # Register Blueprints here
+    from app.task_routes import task_bp
+    app.register_blueprint(task_bp)
+
+    from app.goal_routes import goal_bp
+    app.register_blueprint(goal_bp)
```
Comment (kelsey-steven-ada) [inline app/__init__.py:37]: Nice, I like the choice to divide the routes into their own files!

File: app/models/goal.py [hunk 3]
```diff
@@ -1,5 +1,10 @@
 from app import db
 
-
 class Goal(db.Model):
     goal_id = db.Column(db.Integer, primary_key=True)
+    title = db.Column(db.String, nullable=False)
+    tasks = db.relationship("Task", back_populates="goal", lazy=True)
```
Comment (kelsey-steven-ada) [inline app/models/goal.py:6]: If we check out the SQLAlchemy docs for loading items: https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html the docs say: 
> "The default value of the relationship.lazy argument is "select", which indicates lazy loading." 

Doing some more digging about the possible lazy parameter values we might find further info like in https://medium.com/@ns2586/sqlalchemys-relationship-and-lazy-parameter-4a553257d9ef where we see the line: 
> "lazy = ‘select’ (or True)".

Putting these together we can see that the default value for `lazy` in a db relationship is `select` which is the same as `True`. Since we aren't changing the default value for `lazy`, we could leave this parameter off and our code would behave the same:
```python
tasks = db.relationship("Task", back_populates="goal")
```

File: app/models/task.py [hunk 4]
```diff
@@ -1,5 +1,27 @@
...
+    goal_id = db.Column(db.Integer, ForeignKey('goal.goal_id'), nullable=True)
+    goal = db.relationship("Goal", back_populates="tasks")
+
+    def to_json(self):
+        task_dict = {"id": self.task_id,
+                    "title": self.title,
+                    "description": self.description,
+                    "is_complete": bool(self.completed_at)}
+        if self.goal_id:
+            task_dict["goal_id"] = self.goal_id
```
Comment (kelsey-steven-ada) [inline app/models/task.py:18]: Nice way to handle the optional attribute!

File: app/models/task.py [hunk 5]
```diff
@@ -1,5 +1,27 @@
...
+        return task_dict
+
+    @classmethod
+    def from_json(cls, json):
+        task = Task(title = json["title"],
+                    description = json["description"],
+                    completed_at = json.get("completed_at"),
+                    goal_id = json.get("goal_id"))
+        return task
```
Comment (kelsey-steven-ada) [inline app/models/task.py:27]: Great use of `.get` here to return `None` if the data doesn't exist in the `json` object!

File: app/goal_routes.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+from flask import Blueprint, jsonify, request, make_response, abort
```
Comment (kelsey-steven-ada) [inline app/goal_routes.py:1]: It looks like `abort` and `make_response` aren't being used by this file, so we would want to remove their imports as part of our clean up/refactor steps.

File: app/goal_routes.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    db.session.commit()
+
+    return {"id": goal.goal_id,
+            "task_ids": [task.task_id for task in goal.tasks]}
+
+@goal_bp.route("/<goal_id>/tasks", strict_slashes=False,methods=["GET"])
+def get_goal_tasks(goal_id):
+    goal = validate_id(Goal, goal_id)
+    response = goal.to_json()
+    response["tasks"] = [task.to_json() for task in goal.tasks]
```
Comment (kelsey-steven-ada) [inline app/goal_routes.py:77]: This list comprehension `[task.task_id for task in goal.tasks]` is duplicated in  `add_task_list` and `get_goal_tasks`. How could we update the `Goal.to_json` function to make it flexible enough to include the tasks when desired? Doing so would let us remove the duplicated list comprehensions and use `Goal.to_json` in both places, reducing redundancy and places for possible bugs.

File: app/goal_routes.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+
+@goal_bp.route("", strict_slashes=False, methods=["POST"])
+def create_goal():
+    request_body = request.get_json()
+    
+    try:
+        new_goal = Goal(title=request_body["title"])
+    except KeyError:
+        error_message("Invalid data", 400)
```
Comment (kelsey-steven-ada) [inline app/goal_routes.py:23]: Nice use of an error message when data is missing!

I'd recommend adding explicit info about what key or keys are missing in the response message to help guide the user to what was wrong in their request. The info could be as little as:
```python
error_message("Invalid data: title is missing", 400)
```

File: app/goal_routes.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+from app.models.task import Task
+from app import db
+from app.helper import validate_id, error_message
+
+goal_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+@goal_bp.route("", strict_slashes=False, methods=["GET"])
+def get_goals():
+    goals = Goal.query.all()
+    response = [goal.to_json() for goal in goals]
```
Comment (kelsey-steven-ada) [inline app/goal_routes.py:12]: Great use of list comprehensions across the project!

File: app/goal_routes.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            task = Task.query.get(task_id)
+            task.goal_id = goal_id
+    except KeyError:
+        error_message("Invalid data", 400)
+
+    db.session.commit()
+
+    return {"id": goal.goal_id,
+            "task_ids": [task.task_id for task in goal.tasks]}
+
```
Comment (kelsey-steven-ada) [inline app/goal_routes.py:72]: You're really consistent with whitespace overall which is awesome! Here there's one line between `get_goal_tasks` and `add_task_list`, I'd consider increasing it to 2 lines to be consistent with the spacing between the other functions in the file.

File: app/helper.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+from app.models.task import Task
+from app.models.goal import Goal
+import os
+import requests
+
+def error_message(message, status_code):
+    abort(make_response({"details": message}, status_code))
+
+
+def validate_id(cls, id):
```
Comment (kelsey-steven-ada) [inline app/helper.py:11]: Awesome to see a generic function that can be used with multiple classes! I'd consider naming this something that reflect what the function does, which is returning a record if it is found.

File: app/helper.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    try:
+        id = int(id)
+    except (ValueError, TypeError):
+        error_message(f"invalid id {id}", 400)
```
Comment (kelsey-steven-ada) [inline app/helper.py:15]: If you wanted to include the class name in the error message, we can look it up on an object or class using syntax like:
```python
cls_name = type(cls).__name__
```

File: app/helper.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        error_message(f"id {id} not found", 404)
+    return model
+
+
+# def form_json_response(model):
+#     if type(model) == Task:
+#         response = {"task": model.todict()}
+#     if type(model) == Goal:
+#         response = {"goal": model.todict()}
+#     return jsonify(response)
```
Comment (kelsey-steven-ada) [inline app/helper.py:29]: We want to make sure to delete commented out code as part of clean up. When we are sharing a codebase with other folks, it's unclear what the intention of commented out code is, especially without extra comments explaining why it's there. We use versioning tools like git so we can confidently delete code and view our commit history if we need to recover something we wrote prior.

File: app/helper.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def post_to_slack(task):
```
Comment (kelsey-steven-ada) [inline app/helper.py:32]: Nice helper for the Slack API call!

File: app/task_routes.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+task_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+
+@task_bp.route("", strict_slashes=False, methods=["GET"])
+def get_tasks():
+    sort_order = request.args.get("sort")
+    if sort_order == 'asc':
+        #sort ascending
+        tasks = Task.query.order_by(Task.title.asc())
+    elif sort_order == 'desc':
+        tasks = Task.query.order_by(Task.title.desc())
```
Comment (kelsey-steven-ada) [inline app/task_routes.py:17]: Great use of SQLAlchemy's tools to handle sorting.

File: app/task_routes.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+
+@task_bp.route("", strict_slashes=False, methods=["POST"])
+def create_task():
+    request_body = request.get_json()
+
+    try: 
+        new_task = Task.from_json(request_body)
+    except KeyError:
+        error_message("Invalid data", 400)
```
Comment (kelsey-steven-ada) [inline app/task_routes.py:32]: The feedback in `create_goal` applies here too.

File: app/task_routes.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    db.session.commit()
+
+    response = response = {"task": task.to_json()}
+    return jsonify(response), 200
+
+
+@task_bp.route("/<task_id>", strict_slashes=False, methods=["DELETE"])
+def delete_task(task_id):
+    task = validate_id(Task, task_id)
+    response = {"details": f'Task {task.task_id} "{task.title}" successfully deleted'}
```
Comment (kelsey-steven-ada) [inline app/task_routes.py:65]: Nice, I like the informative message here!

File: app/task_routes.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    return jsonify(response), 200
+
+
+@task_bp.route("/<task_id>", strict_slashes=False, methods=["PUT"])
+def update_task(task_id):
+    request_body = request.get_json()
+    task = validate_id(Task, task_id)
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+    task.completed_at = request_body.get("completed_at")
```
Comment (kelsey-steven-ada) [inline app/task_routes.py:54]: When we get a KeyError trying to creating an object in `create_task` we catch the error and return some info to the user rather than letting the default 500 error without troubleshooting info pass through. It would be helpful for users to do something similar here as well.

File: tests/test_wave_01.py [hunk 19]
```diff
@@ -60,13 +60,23 @@ def test_get_task_not_found(client):
     # Assert
     assert response.status_code == 404
 
-    raise Exception("Complete test with assertion about response body")
+    #raise Exception("Complete test with assertion about response body")
     # *****************************************************************
     # **Complete test with assertion about response body***************
     # *****************************************************************
+    assert response_body["details"] == "id 1 not found"
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:67]: To ensure we're checking all the relevant data, it might be good to confirm that the database has no records, or that there is no record in the DB with an ID of 1. (This feedback around checking the DB applies to other tests as well.)

File: tests/test_wave_01.py [hunk 20]
```diff
@@ -60,13 +60,23 @@ def test_get_task_not_found(client):
...
 
 
-@pytest.mark.skip(reason="No way to test this feature yet")
+def test_get_task_id_not_numeric(client, one_task):
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:70]: Nice! Awesome to see you found edge cases that aren't covered and made your own tests 😄

File: tests/test_wave_05.py [hunk 21]
```diff
@@ -80,34 +81,44 @@ def test_create_goal(client):
...
     # ---- Complete Act Here ----
-
+    response = client.put("/goals/1", json={"title": "Brand New Title"})
+    response_body = response.get_json()
     # Assert
     # ---- Complete Assertions Here ----
     # assertion 1 goes here
+    assert response.status_code == 200
     # assertion 2 goes here
+    assert "goal" in response_body
```
Comment (kelsey-steven-ada) [inline tests/test_wave_05.py:96]: Since we check the whole structure of `response_body` on the line below, we could omit this specific check for `goal` in `response_body`. We'd still want a 3rd assertion, what else would be helpful to look at here?

**Review by kelsey-steven-ada — COMMENTED** [review]
Great work pulling together many concepts in this project Danielle! Please take a look at the feedback when you have time, and reach out if you have questions or could use clarifications on anything 😊
